### ISSUE TITLE
ENOMEM error with micropython 1.20.0

### ISSUE THREAD

**Issue by tux-friend** [Marcelhag/phyphox-mpy#20]
When using the ,CreateExperiment.py' example with Micropython 1.20.0 the following error occurs.

```
Traceback (most recent call last):
  File "<stdin>", line 122, in <module>
  File "<stdin>", line 110, in main
  File "phyphoxBLE/phyphoxBLE.py", line 105, in write
OSError: [Errno 12] ENOMEM
>Unhandled exception in IRQ callback handler
Traceback (most recent call last):
  File "phyphoxBLE/phyphoxBLE.py", line 92, in _irq
  File "phyphoxBLE/phyphoxBLE.py", line 204, in when_subscription_received
OSError: [Errno 12] ENOMEM
```

**Comment by Staacks** [comment]
Thanks for reporting this. In fact, we just stumbled upon this problem on Monday when all students in our lab course suddenly failed to use the phyphoxBLE module.

The short answer is that for now you will have to stick to v1.19.1, which still works perfectly fine.

The long answer is that so far I have tried a few things to avoid handing newly allocated memory to the notify function and removing BytesIO since I suspected that either kept a reference leading to a memory leak. However, even repeatedly sending the same buffer eventually ends in an ENOMEM and I am beginning to suspect a bug in 1.20.0. So, my next steps will be creating a minimal example and see if I can reproduce the problem this way to pinpoint it to MicroPython.

**Comment by tux-friend** [comment]
Ok, thanks! I suspected that it might be a bug of 1.20.0. other question, I did a write up on how to install phyphoxBLE with Thonny, should I pull the request on this repository also?

Ah and thanks for the great work! I really like phyphoxBLE as it is very easy and useful for teaching physics!

**Comment by Staacks** [comment]
> Ok, thanks! I suspected that it might be a bug of 1.20.0.

Sorry, I found the reason and it is our fault after all. The experiment transfer blocks the interrupt function of the BLE module, which 1.20.0 does not seem to like. I am not entirely sure if this is truly directly called from an interrupt handler or rather a callback, which should be able to deal with a function that blocks for a while (like 1.19.1 did), but it is definitely bad style at our end and something we can fix by cleaning up our code. So, I am reopening this for now until I get to to this in the next days.

> other question, I did a write up on how to install phyphoxBLE with Thonny, should I pull the request on this repository also?

I just realize that this is Marcel's repository and that the official repository is still a fork of this one. Marcel is one of the student assistants who developed the MicroPython library (and who is currently also working on the new web editor). But eventually he will graduate and leave the project, so we will detach the official repo while this one here will become stale (unless Marcel keeps merging things back).

Long story short: Please send the PR to https://github.com/phyphox/phyphox-micropython and I will also push the fix there (with Marcel probably merging things back here to close the issue).

Did you end up in this repository here because it is currently still referenced as the original one or is there an obsolete link somewhere that we need to redirect?

**Comment by tux-friend** [comment]
Ok, great! Yes, I ended up in this repository because it is currently referenced as the original.

**Comment by Staacks** [comment]
The fix is now in https://github.com/phyphox/phyphox-micropython and as version 0.1.1 available on PyPI.
I have also detached (apparently the right term is "extracted") the other branch from this one here.

I think that should be all :)
